### ISSUE TITLE
Atlas: Check Organization during Oauth

### ISSUE THREAD

**Issue by LindseySaari** [department-of-veterans-affairs/va.gov-team#72771]

**Comment by RachalCassity** [comment]
New Oauth App is being tested :)

**Comment by jennb33** [comment]
Need to determine who would approve this - Bill or Clint, TBD. @LindseySaari can help to get this completed.

**Comment by jennb33** [comment]
Per @RachalCassity - we can get the organization from OAuth, need to do some discovery around that.

**Comment by LindseySaari** [comment]
We will also have to be mindful of users being removed from the organization and invalidate their sessions

**Comment by LindseySaari** [comment]
## Notes update:

The Github apps have been created. To be clear, this is a Github app not an Oauth app ([see the differences detailed here](https://docs.github.com/en/apps/creating-github-apps/about-creating-github-apps/deciding-when-to-build-a-github-app)). Github apps allow for more granular scopes/permissions. 

The goal here is to integrate an organization check [upon the auth callback](https://github.com/department-of-veterans-affairs/platform-atlas/blob/master/app/controllers/sessions_controller.rb#L8) to ensure that the user trying to authenticate is in the department-of-veterans-affairs Github org. For Oauth, we use the [Github omniauth gem](https://github.com/omniauth/omniauth-github), which is just an Oauth wrapper. 

We've already created the Github apps (dev and prod Atlas), but we need to adjust some of the organizational permissions in order to obtain the member information (see [creating a Github app page](https://github.com/settings/apps/new) and expand the permission sections). The permissions of a GitHub App determine what the app can do with GitHub's APIs. When a user authorizes the app to act on their behalf, the GitHub App can use the resulting user access token to make API requests. Organization permissions allow your app to access resources related to the organization where the app is installed, if it is installed on an organization account. We have [a request started here](https://dsva.slack.com/archives/CU1E4CX9U/p1712076501013629) to obtain the approval or management role to adjust the Github app permissions.

Essentially, following this approval, we should be able to adjust the Github app permissions. Upon doing so, we can use the users access_token returned on the callback to make a request to the Github API to see if the members is in the organization.

Some things to consider. 
1. We will need to [adjust the scope](https://github.com/department-of-veterans-affairs/platform-atlas/blob/master/config/initializers/omniauth.rb#L20) to [read:org](https://github.com/omniauth/omniauth-github?tab=readme-ov-file#scopes)
2. We will want to extend the [from_omniauth method](https://github.com/department-of-veterans-affairs/platform-atlas/blob/master/app/models/user.rb) to do a check on the organization
3. Users can be removed from the org at any time. We want to prevent unauthorized access. To do so, we will want to to do a check 
4. Storing and refreshing access tokens.

```
class ApplicationController < ActionController::Base
  # ...

  def authorize_user!
    return if auth_not_required_for_path? || (current_user.present? && current_user.is_member?)

    flash[:error] = 'Please sign in.'
    redirect_to(root_path)
  end
end

class User
  # ...

  def is_member?
    Rails.cache.fetch("#{@user.github_username}/is_member", expires_in: 12.hours) do
      @octo_client = Octokit::Client.new(access_token: token_here) ## Oktokit call
      @octo_client.organization_member?('department-of-veterans-affairs', user.github_username)
    end
  end
end
```

**Comment by LindseySaari** [comment]
I've created [an issue here](https://github.com/department-of-veterans-affairs/github-user-requests/issues/21221) to  get some help from a Github Admin so that we can get permissions to adjust the organizational setting on the Github apps for Atlas

**Comment by RachalCassity** [comment]
Waiting for approval. I will test locally after the GithubApps are approved.

**Comment by rmtolmach** [comment]
Still waiting for approval, but Lindsey's branch is here: https://github.com/department-of-veterans-affairs/platform-atlas/compare/master...github-organization-check

**Comment by LindseySaari** [comment]
We are blocked on this until Github allows us to move forward

**Comment by jennb33** [comment]
@jennb33 to This is the Github issue to keep an eye on. https://github.com/department-of-veterans-affairs/github-user-requests/issues/21221

You could even make a comment tomorrow (4/11) or Friday (4/12)  to nudge the GH folks

**Comment by jennb33** [comment]
Pinged GitHub admins on 4/11 pm in [21221](https://github.com/department-of-veterans-affairs/github-user-requests/issues/21221)

**Comment by jennb33** [comment]
We have a response in 21221; they need to update their documentation tho. Requested that documentation update on 4/15/2024

**Comment by LindseySaari** [comment]
The github apps are still not approved but we've found a workaround with the Oauth apps (which are more basic) than using the Github app for now. The basic Oauth app doesn't provide the ability for a refresh token, so essentially the user will just get logged out instead of calling for a refresh token. Once we get the Github app implemented, we can move foward with refresh token functionality.

**Comment by LindseySaari** [comment]
Our Github apps have been [approved](https://github.com/department-of-veterans-affairs/github-user-requests/issues/21221#issuecomment-2065120948)! This will block the future work of the automated refresh of the access token for future Atlas work in #80919
